Support changing the tree display format

rowo360 opened the pull request:
This PR is related to issue #1101. However it's not completed yet, but it's intended to check that I'm on the right track. So it's more a preview about the upcoming changes - for example all tests are missing. But this commit already contains more changes than I did before, so some confirmation is helpful.

For a list of use cases / test cases for this issue see my comment in #1101.

The essential part from a technical point of view is that I introduced a new setting 'DisplayFormat'. So whenever this setting changes, the TreeViewPresenter can update the tree and the TestCentricPresenter can update the selected menu items. I copied this approach from the GroupBy functionality - which also use a setting as the central store and notification.
I hope this is aligned with the general design.

While implementing/testing these changes, these considerations came into my mind:

- Should we align the wording: sometimes it's called 'DisplayFormat' and sometimes 'DisplayStrategy'? I would be a good point in time to harmonize this wording - I would prefer 'DisplayStrategy' as all those classes are also named XYZStrategy. But we could keep it unchanged of course also.
- I wonder if I should try to introduce an interface IStrategy for testability reasons. Currently the class TreeViewPresenter is working with the concrete classes DisplayStrategy/NUnitTreeDisplayStrategy/... which makes testing a little bit harder.

Thanks for your feedback :-)

CharliePoole commented:
I looked at the code but I won't yet do a line-by-line comment. I think it does what you are trying to do but we need to first discuss the "what", in particular as applied to the difference between global settings those currently applied to a particular test project.

Regarding your two questions...

* I think that the _user_ wants to see a particular Display Format. "Strategy" is a design pattern and even though our users are typically developers, I don't think it belongs in a user interface. Internally, I'd say that the code related to the view should use "format", while the code in the presenter should continue to use "strategy". Note that if you include groupings there are lots more formats than strategies.
* Great idea! I think that `IStrategy` is too generic though. Maybe ``ITreeDisplayStrategy`? If you do this, I'd prefer to see it first, as a separate PR labeled "Refactor."

I'll move over to the issue to document the "what we should do" part.

rowo360 commented:
Great, I was hoping for exactly this discussion or exchange, because it became a bit more complex than I had initially assumed.
Regarding the idea about introducing the ITreeDisplayStrategy interface: I'll take a closer look and see if it's feasible for me. Meanwhile we can continue to discuss the use cases.

CharliePoole reviewed (approved):
Based on our discussion, this looks good to me. Not merging as you said it's incomplete.

rowo360 commented:
I have continued to work on this task: basically filling some gaps in the implementation and extending the tests. Overall, I have the impression that the implementation for this issue is already in place and that I only had to fill in a few gaps. so that everything works together. And I consider this as good news :-)

So finally I' ve good feeling that users can work with the different tree display formats, now. Unfortunately it's not perfect yet - I noticed crashes in two scenarios. However I prefer to create separate issues (+ PRs) for those - because maybe it needs to be discussed in detail first and it would enlarge this PR further.

Some explanations about the last commit:
**TreeViewPresenter.cs**
- Event.RunStarting: we discussed and agreed upon #1094 to clear the previous test results before starting a new test run. But unfortunately I missed to clear the results also in the Model! I have now caught up on that. I only noticed this error when changing the tree DisplayFormat.
- Settings changed DisplayFormat/GroupBy: I noticed that the Strategy is created/Reload occassionaly twice, when loading a test project. That's not optimal and I want to avoid it. (I expect some performance impact for large test projects) Therefore I added some statements which checks if the Strategy is already up-to-date. 
How can this happen: while loading a test project the Strategy is created, but we also update the settings from the VisualState file. So this settings might invoke an event, which is handled by this class and would like to update the strategy.
- RunContextCommand: This change is for the 'Run' context-menu handler : so far nothing happened for grouped nodes. Now all test cases of that group are executed.

**FixtureListDisplayStrategy/TestListDisplayStrategy**
- The new method IsGroupingUpToDate() checks if the current grouping is up-to-date with the grouping settings. 

**DisplayStrategy**
- Method ApplyResultToGroup: I added a null check here, because I ran into crashes regularly. It might be the case that some further implementation is required here to fulfill all use cases, but I didn't notice any obvious gap right away.

**TestCentricPresenter**
- I added a statement in the constructor so that the DisplayFormat context-menu is initialized properly from the current settings (SelectedItem + Enable/Disable)

Ok, that's all - sorry, a lot of text to read, but I hope it helps to understand the commit better.
If it's OK, I think we should merge PR #1103  first, then I update this branch and finally this PR can be merged?

rowo360 commented:
@CharliePoole:
Unfortunately, I now notice that the pipeline is displaying an error: Microsoft.NETCore.App', version '5.0.0' is missing for some reasons. I doubt that my commit causes this error but instead something changed on the server? At least I can see in the log files that the .Net SDK version and the MSBuild version has slightly changed. 
Any idea how to fix this one?

rowo360 commented:
Perfect, that was the solution!
I also found a note about the image update, now (https://www.appveyor.com/updates/)
I think I've moved myself not into the best situation by having two PRs open with conflicting changes. But that's also a learning! :-)
So I'll commit this appveyor fix also to that other PR now, with the intention that it get merged first and I solve the conflicts afterwards - Ok?

CharliePoole commented:
Sounds good. Also, think about starting to use the TestCentric org for each thing you work on rather than your fork. I can coach you through doing that if you need help.

rowo360 commented:
@CharliePoole :
I updated this PR with the lastest changes from main and think that it's now ready for merge so that it closes #1101 .
The changes from main bring along the new ITreeDisplayStrategy interface, so that I need to update some of the new tests.
I'll use Testcentric org instead of my fork next time when starting with a new issue.

CharliePoole commented:
@rowo360 Slightly off-topic, but...

Reviewing this I noticed that a stylistic choice I made in the past created some problems for reviewers. If you play reviewer to your own code you may notice this as well.

In most presenters I have a "WireUpEvents" method in which I create anonymous methods using lambdas for each event. I felt this made reading the code cleaner, because everything is in one place and you don't have to scroll down to see the actual methods. 

But, when reviewing changes on GitHub, the display is organized according to __actual__ methods, and so I need to do some clicking and expanding and scrolling to figure out the context.

I don't want to change anything now, but it's interesting that my approach, grounded in the notion that the code is only viewed in an editor, falls down when looking at it thorough the lense of a GitHub PR.

However, any change is for the future, not for this PR. :-) Take a look and see what you think though.

CharliePoole reviewed (changes requested):
Looks great! Nice tests especially. Only a few minor suggestions for change plus some ideas for the future to think about.

rowo360 commented on src/TestCentric/testcentric.gui/Presenters/TreeViewPresenter.cs:
```diff
@@ -148,6 +164,9 @@ private void WireUpEvents()
                         _model.RunTests(testNode);
+                    var groupNode = _view.ContextNode.Tag as TestGroup;
+                    if (groupNode != null)
+                        _model.RunTests(groupNode);
```
Ok, I'll adapt this.

rowo360 commented on src/TestCentric/testcentric.gui/Presenters/TreeViewPresenter.cs:
```diff
@@ -112,9 +115,22 @@ private void WireUpEvents()
+                                Strategy.Reload();
+                            }
+                            break;
+                        }
```
I think it's good that you point out this code again. As a result, I looked at this again, checked it and debugged it. And finally I came to the conclusion that I prefer to remove this check! 
It's not that my intention was wrong, but it doesn't really seem necessary to me now and therefore I'd rather just leave it simple at that. My original observation was that the Strategy might be created/reloaded twice during project loading and that's still true. And I observed that during Strategy creating/reloading all tree nodes are created. At that point of time I stopped analyzing and started to implement that check to prevent that double loading. Overall I was worried that this double loading will cause some performance impact.
But based on your input (the UI is disabled during loading), I checked my concerns with an example. And as a result, I did not notice any deterioration in performance (I used a sample project with ~5000 test cases).

Didn't someone famous say a long time ago that you shouldn't optimize without measuring? ;-)
I would have listened better :-)

So, I'll rollback these lines and also the next change!

CharliePoole replied:
I think the check would  be correct if this were an independent presenter designed to work with multiple view implementations. When I started to experiment with MVP I had that idea but in the end concluded it was too complex for my purposes. So now the view and presenter are in a one to one relationship and the only purpose of the interface is to facilitate testing. What will remain after this is to better separate the testing of the presenter and each strategy, by use of the interface you added.

rowo360 commented on src/TestCentric/testcentric.gui/Presenters/TestCentricPresenter.cs:
```diff
@@ -923,7 +937,9 @@ private void SetTreeDisplayFormat(string displayFormat)
-                    ((Elements.CheckedToolStripMenuGroup)_view.GroupBy).MenuItems[1].Enabled = false;
+                    Elements.CheckedToolStripMenuGroup menuGroup = _view.GroupBy as Elements.CheckedToolStripMenuGroup;
+                    if (menuGroup != null)
+                        menuGroup.MenuItems[1].Enabled = false;
```
I also read the HACK comment and spend some thoughts how to fix this best, but didn't come up with a solution. Do you some hint, here?

rowo360 replied:
Unfortunately the statement `menugroup?.MenuItems[1].Enabled = false;` is not accepted by the c# compiler (something about 'left side of an assignment...'), so we need to stick with the lengthy statement.

CharliePoole replied:
Ah. I'm not a good substitute for a compiler then. :-)

rowo360 commented:
@CharliePoole :
May I ask for your help regarding the merge? I would like to start the merge, but it is currently blocked. And it seems that there's one change requested which I did not addressed yet. But I cannot figure out which one and if it's a mandatory change. I had the impression that it's about the `menugroup?.MenuItems[1].Enabled = false;` comment and tried to address it with one additional commit but without success. I also tried to set all those review comments to 'resolve converation', but it doesn't have an effect.
So overall, I'm struggeling with GitHub what's required to be done next or maybe I cannot do it myself?

CharliePoole commented:
@rowo360 OK... it wanted me to approve the changes you made. "One" reviewer had requested changes and needed to be approved. IMO GitHub is imposing more regulation in the process than I wanted to see. I prefer to be able to work on the basis of trust when you are made a committer. But we'll see if it acts differently when you are no longer using a fork.

So, as there is no rush to merge this, please continue the experiment and try to merge it again. It's useful knowledge for improving our scripts as well.

rowo360 commented:
Yes, of course - let's keep trying that I merge this PR.
However it seems that some basics rights are still missing, as I cannot start the merge process (missing write access).

![grafik](https://github.com/user-attachments/assets/82c73087-2cf9-4b21-8752-1845c969ea1e)

CharliePoole commented:
OK, I hadn't given the gui-team write access. Try again.

CharliePoole commented:
That did it then. Both GitHub and AppVeyor are OK with you. :-) We now have a new dev release on myget, 2.0.0-dev00451, which happens automatically whenever a PR is merged to main. You should either close #1101 or add a comment about what's left to do at this point.

rowo360 commented:
Perfect - I'm glad that it worked! :-)
I started the merge this morning and got immediately feedback that it was successfully. But although I was happy, I was wondering what exactly 'successful' means and what's happening behind the scenes.
Now, I can see in Github that I merged the latest PR and there's also a link to the Appveyor run. Besides that, there's no other link/log/test run on GitHub itself, isn't it?
I'll have a closer look to the cake build script again. I think it is described there what should be carried out during a merge operation. Also that part with the myget package. Do you have a link to that one?
I assume it's exactly the intention that it's not public, right?

CharliePoole commented:
Our MyGet feed is publicly available using the nuget and choco commands or by including the feed in your nuget.config. It also has an IDE, which requires signin. I'll send you some info by email about how we use it and direct others to it.
